### DISCUSSION TITLE
Add simple image edit/mask component

### DISCUSSION THREAD

**Discussion by iwr-redmond** [rio-labs/rio#168] [Feature requests]
Much ML work involves image generation and refining, which means that a component capable of simple edits (e.g. crop) and especially adding an inpainting mask could be very helpful.

Some libraries good enough for government work:

- [Mini Canvas Editor](https://github.com/img-js/mini-canvas-editor): lightweight and can be customized for specific use cases
- [TUI Image Editor:](https://github.com/nhn/tui.image-editor): customizable interface with all the various features
- [Pikaso](https://github.com/pikasojs/pikaso): can have a background image to paint on a mask, and separately has a crop function

**Comment by iwr-redmond** [comment]
Looking at the [Mini Canvas Editor configuration file](https://github.com/img-js/mini-canvas-editor/blob/main/editor/src/editor-configuration.ts), I think this would be best.

Demos of the features I suggested might be a priority for ML apps:

- [Inpainting](https://img-js.github.io/mini-canvas-editor/webpack-app/public/inpainting-mask.html): [config file](https://github.com/img-js/mini-canvas-editor/blob/main/demos/webpack-app/src/inpainting-mask.ts)
- [Cropping](https://img-js.github.io/mini-canvas-editor/webpack-app/public/crop.html): [config file](https://github.com/img-js/mini-canvas-editor/blob/main/demos/webpack-app/src/crop.ts)

**Comment by iwr-redmond** [comment]
**Inpainting**

Each line drawn is a separate layer on the canvas. The [layer-item.ts](https://github.com/img-js/mini-canvas-editor/blob/main/editor/src/sidebar/layers/layer-item.ts) code has at L100:

```
private readonly onDeleteClicked = () => {
		this.state.canvas.remove(this.object);
	};
```

Suggesting that all of the markers could be deleted with a forEach loop similar to [template-creator.ts L88](https://github.com/img-js/mini-canvas-editor/blob/main/demos/webpack-app/src/template-creator.ts). I suppose this could also allow the first click to be recognized and used to activate Segment Anything in the future.

The [configuration file](https://github.com/img-js/mini-canvas-editor/blob/main/editor/src/editor-configuration.ts) has code at L28:

```
export interface BrushModeConfiguration {
	brushSize?: number;
	brushColor?: string;
}
```

Allowing easy theme integration.

**Cropping**

The demo resizes the image on the canvas rather than cropping. However, "crop" might be better expressed in the [Template Creator example](https://img-js.github.io/mini-canvas-editor/webpack-app/public/template-creator.html), with the white box representing the fixed canvas and the green $cat box representing the crop mask.

Per the backend code [template-creator.ts](https://github.com/img-js/mini-canvas-editor/blob/main/demos/webpack-app/src/template-creator.ts), this would look something like:

```
const editor = Editor.createFromImage(
			placeholder,
			image,
			{
				selectable: false
			},
editor.add(
	new MceRect({
		fill: 'transparent_png_file',
		left: 0,
		top: 0,
		width: 100%,
		height: 100%,
		label: 'Mask'
	})
);

```
Bearing in mind that my code is good enough for government work.

**Comment by mad-moo** [comment]
I am confliced. As you've said, components like this are currently very in demand. The thing is, they're highly specialized for this exact subfield of AI, and given how rapidly AI is moving, the component probably won't be used for long, or would at least have to undergo massive changes.

This would be a perfect candidate for a user defined HTML component. Something that doesn't ship with Rio, but has access to the HTML DOM & JavaScript. ***Except***, we don't have those yet, at least not publicly. If you need this urgently I can help you create your own HTML component using the internal, not-yet-stable API. If not, I think this should wait for that API to come online, at which point it could live in a separate python package. What do you think?

**Comment by iwr-redmond** [comment]
I think the right path depends on the complexity of integration. If the complexity level is low, then a custom component would make sense; if high, forget it. However, I suspect the complexity level is moderate, which makes me think that an image editor function makes most sense in core.

In terms of long-term need, beyond current popularity, revenue from AI image generation is predicted to grow by [17.7% (compound) each year to 2030](https://www.grandviewresearch.com/industry-analysis/artificial-intelligence-ai-image-generator-market-report).

**Comment by iwr-redmond** [comment]
I suggest adding [Jimp](https://github.com/jimp-dev/jimp) to the list of possible editor components. It's lower level than Mini Canvas Editor, which means it would probably be more work to implement initially but might provide more long term flexibility. Customization options are documented [here](https://github.com/jimp-dev/jimp/blob/main/packages/docs/src/content/docs/guides/custom-jimp.mdx).

A few more data points supporting this feature request:

According to a recent paper by Zhang and Tang, ML image research is growing rapidly - see [Table 3 (p9)](https://arxiv.org/pdf/2411.16164#page=9). ([Table 4](https://arxiv.org/pdf/2411.16164#page=11) paints a similar picture, so to speak, but is not sorted by year.)

Black Forest Labs have recently released a model dedicated to [inpainting and outpainting](https://huggingface.co/black-forest-labs/FLUX.1-Fill-dev), giving it pride of place in the [press release](https://blackforestlabs.ai/flux-1-tools/).

Inpainting research goes back over two decades and is so common that there have been at least 45 reviews summarizing earlier papers, according to the October 2024 special issue of the Computers & Graphics journal ([Barcelos, Rabelo, Bernadini, et al, 2024](https://www.sciencedirect.com/science/article/abs/pii/S0097849324001456)).

**Reply by iwr-redmond** [reply]
Some numbers and analysis from the recent paper [by Xia, Zhu, Li, et al](https://arxiv.org/pdf/2504.08619):

- Computer vision topics currently  account for ~10% of papers
- Vision-related papers were in the top 10 topics for 1/3 of conferences
- Multi-modal and vision papers have "rising traction"
